trieste/models/keras/architectures.py

```diff
@@ -0,0 +1,315 @@
+
+        return tf.keras.Model(inputs=inputs, outputs=outputs)
+
+
+class KerasEnsembleNetwork(tf.Module):
+    """
+    This class is an interface that defines necessary attributes and methods for neural networks
+    that are meant to be used for building ensembles by
+    :class:`~trieste.models.keras_networks.KerasEnsemble`. Subclasses are not meant to
+    build and compile Keras models, instead they are providing specification that
+    :class:`~trieste.models.keras_networks.KerasEnsemble` will use to build the Keras model.
+    """
+
+    def __init__(
+        self,
+        input_tensor_spec: tf.TensorSpec,
+        output_tensor_spec: tf.TensorSpec,
+        network_name: str = "",
+    ):
+        """
+        :param input_tensor_spec: Tensor specification for the input to the network.
+        :param output_tensor_spec: Tensor specification for the output of the network.
+        :param network_name: The name to be used when building the network.
+        """
+        if not isinstance(input_tensor_spec, tf.TensorSpec):
+            raise ValueError(
+                f"input_tensor_spec must be an instance of tf.TensorSpec, "
+                f"received {type(input_tensor_spec)} instead."
+            )
+        if not isinstance(output_tensor_spec, tf.TensorSpec):
+            raise ValueError(
+                f"output_tensor_spec must be an instance of tf.TensorSpec, "
+                f"received {type(output_tensor_spec)} instead."
+            )
+
+        self.input_tensor_spec = input_tensor_spec
+        self.output_tensor_spec = output_tensor_spec
+        self.network_name = network_name
+
+    @property
+    def input_layer_name(self) -> str:
+        return self.network_name + "input"
+
+    @property
+    def output_layer_name(self) -> str:
+        return self.network_name + "output"
+
+    @property
+    def flattened_output_shape(self) -> int:
+        return int(np.prod(self.output_tensor_spec.shape))
+
+    @abstractmethod
+    def connect_layers(self) -> tuple[tf.Tensor, tf.Tensor]:
+        """
+        Connects the layers of the neural network. Architecture, layers and layer specifications
+        need to be defined by the subclasses.
+
+        :return: Input and output tensor of the network, required by :class:`tf.keras.Model` to
+            build a model.
+        """
+        raise NotImplementedError
+
+
+class GaussianNetwork(KerasEnsembleNetwork):
+    """
+    This class defines layers of a probabilistic neural network using Keras, with Gaussian
+    distribution as an output. The network architecture is a multilayer fully-connected
+    feed-forward network. The network has to be built and compiled as an ensemble model,
+    it should also be trained with maximum likelihood loss function.
+    """
+
+    def __init__(
+        self,
+        input_tensor_spec: tf.TensorSpec,
+        output_tensor_spec: tf.TensorSpec,
+        hidden_layer_args: Optional[List[dict[str, Any]]] = None,
```

Comment:
By allowing sequences you can also make the default explicit:

```suggestion
        hidden_layer_args: Sequence[dict[str, Any]] = (
            {"units": 50, "activation": "relu"},
            {"units": 50, "activation": "relu"},
        ),
```